### ISSUE TITLE
Smearing issue with TransformerLayer

### ISSUE THREAD

**Issue by hjweide** [Lasagne/Lasagne#689]
It seems to me as if the pixels that lie on the boundary of the input to the TransformerLayer are smeared in the output.  I noticed this when I was playing around with spatial transformers by transforming MNIST digits and trying to let the transformer module learn to map the digit back to its original pose.

In the image below, the first row shows the input, the second row shows the output of the spatial transformer, and the third row shows the output of the spatial transformer after the boundary of the input has been set to zero.  Notice that in the second row smearing occurs wherever the input digit was touching the boundary, whereas the third row does not exhibit this effect.
![smearing](https://cloud.githubusercontent.com/assets/6610704/15592112/145fe6c6-236f-11e6-9168-f484a225141d.png)

I suspect that the issue arises when the coordinates are [clipped](https://github.com/Lasagne/Lasagne/blob/8fe645d28b66f991d547e9b6a314251b8e84446a/lasagne/layers/special.py#L470), because this leads to several locations in the output grid all mapping back to the same boundary location in the  input.

Here is some code to reproduce the issue.  You can get a sample of transformed MNIST digits [here](https://www.dropbox.com/s/ayxt5jkxtvekri3/transformed_mnist.pickle?dl=0).

```
import cv2
import cPickle as pickle
import numpy as np
import theano
import theano.tensor as T
from lasagne.layers import get_output
from lasagne.layers import DenseLayer
from lasagne.layers import InputLayer
from lasagne.layers import TransformerLayer
from lasagne.init import Constant
from lasagne.nonlinearities import linear

with open('transformed_mnist.pickle', 'rb') as f:
    X = pickle.load(f)
Y = np.copy(X)
Y[:, :, 0, :] = 0
Y[:, :, :, 0] = 0
Y[:, :, -1, :] = 0
Y[:, :, :, -1] = 0

l_in = InputLayer((None, 1, 28, 28))

M = np.array([[1.5, 0, 0.0], [0, 1.5, 0.0]])
l_loc = DenseLayer(
    l_in, num_units=6, W=Constant(0.0), b=M.flatten(), nonlinearity=linear
)
l_trans = TransformerLayer(l_in, l_loc)

X_tensor = T.tensor4()
X_hat = get_output(l_trans, X_tensor)
infer = theano.function([X_tensor], X_hat)

output1 = infer(X)
output2 = infer(Y)
row1 = (255. * np.hstack(X.transpose(0, 2, 3, 1)))
row2 = (255. * np.hstack(output1.transpose(0, 2, 3, 1)))
row3 = (255. * np.hstack(output2.transpose(0, 2, 3, 1)))
cv2.imwrite('smearing.png', np.vstack((row1, row2, row3)))
```

**Comment by f0k** [comment]
Thanks for the report! Unfortunately, I'm not really sure if and what to do about it -- in the case of MNIST digits, replacing out-of-sample pixels with zero is an obvious solution because the background is black, for other images this may not be suitable at all. We could add a border mode argument to the layer to specify what should happen with out-of-sample pixels (clip to border as we do now, fill with a given constant, mirror the input, or repeat the input). `scipy.ndimage` has this for many of its filters:
```
mode : {'reflect', 'constant', 'nearest', 'mirror', 'wrap'}, optional
    The `mode` parameter determines how the array borders are
    handled, where `cval` is the value when mode is equal to
    'constant'. Default is 'reflect'
cval : scalar, optional
    Value to fill past edges of input if `mode` is 'constant'. Default
    is 0.0
```
Mirroring and wrapping are trivial modifications of the current code (replacing `T.clip` with other operations), replacing by a constant is more involved. Maybe it's actually a good workaround to pad the image with a border of the value you'd like. In any case, a PR adding a `border_mode` argument or similarly would be welcome!

**Comment by hjweide** [comment]
I actually completely forgot about this until I came across the similar issue #782.  I've quickly added `mirror` and `wrap` as border modes.  I suspect that `constant` is the most useful one, but I'm not sure how I would implement that.  Let me know if it looks good and I'll make a pull request:
https://github.com/hjweide/Lasagne/blob/master/lasagne/layers/special.py#L476

**Comment by f0k** [comment]
> I've quickly added mirror and wrap as border modes. [...] Let me know if it looks good and I'll make a pull request

Cool! Sorry for the late reply, this looks good at first glance! We should just take care to mimic the scipy conventions, if possible. So it should be `nearest`, not `clip`, and there's a fine distinction between `mirror` and `reflect` (https://mail.scipy.org/pipermail/scipy-user/2008-November/018611.html) we should get right (although we don't need both).

**Comment by hjweide** [comment]
Thanks for pointing out the subtle difference between `mirror` and `reflect`.  I've updated the code to be consistent with scipy.  I've also changed `clip` to `nearest`, but this changes the default value of `TPSTransformerLayer`'s `__init__`, so has the potential to break people's code.  It should be an easy fix for them, but I'll let you decide if we should make a note of this.

If everything looks good I'll submit a pull request:
https://github.com/hjweide/Lasagne/blob/master/lasagne/layers/special.py#L476
